### PR TITLE
[in_app_purchase] improve readme

### PR DESCRIPTION
**Pull request by cyanglaz**
This improves the readme of the plugin.
1. Added missing feature in readme
2. Added the preparation section header
3. Link users to contribution guide
4. Add more code samples
5. Fix some typos  

Upcoming in a separate PR: video demos for the plugin.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I signed the [CLA].
- [ ] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/in_app_purchase/README.md
Comment (jayoung-lee) [inline packages/in_app_purchase/README.md:255]: What about adding a `## Contributing to this plugin` heading for these last 2 paragraphs?

File: packages/in_app_purchase/README.md
Comment (jayoung-lee) [inline packages/in_app_purchase/README.md:245]: I think this section could be a part of the previous section, with a title like `After editing serialized data structs`. (What the current heading `## Development` means wasn't very clear, until I read the whole paragraph. )

File: packages/in_app_purchase/README.md
Comment (jayoung-lee) [inline packages/in_app_purchase/README.md:19]: I suggested adding `### Preparation` here, but another good option would be adding `## Usage` around line 51 to separate the steps for preparation and steps for actually using the plugin.

Comment (cyanglaz) [reply packages/in_app_purchase/README.md:19]: Removed the ###preparation, added ##Usage.

File: packages/in_app_purchase/README.md
Comment (cyanglaz) [reply packages/in_app_purchase/README.md:255]: done

File: packages/in_app_purchase/README.md
Comment (cyanglaz) [reply packages/in_app_purchase/README.md:245]: So this is actually different from the previous section. This section is more for people who contribute or develop the plugin, the previous section is more for people who uses the plugin. Maybe this can also be a part of ## Contributing to this plugin?

**Comment by kwalrath** [comment]
Before I can do a detailed review of this README, I need some info:

* What should the main sections be?
  * It kind of seems like **Preparation** should be a section before **Getting Started**, maybe named **API overview**.
  * Should we rename **Getting Started** to something like **Example code**? (Btw, I love that you have so much example code!)
* The purpose of the **Development** section is unclear.
  * Who is the first paragraph aimed at? A user of the API or a contributor to the package?
  * The second paragraph (about this being a beta plugin) seems pretty important. Should it be up at the top of the README?
* We need to establish some terminology. (It's fine to mention all the terms people might expect, but after the main term is established we should stick with it.) Which term do you want to use from the following groups?
  * _in-app purchase_ or _IAP_?
  * _storefront_, _underlying shop_, or _underlying store_?

**Comment by cyanglaz** [comment]
> * What should the main sections be?
>   
>   * It kind of seems like **Preparation** should be a section before **Getting Started**, maybe named **API overview**.

We don't have a **Preparation** section in the main README. It's only in the example app README, the purpose is to teach users how to setup and run the example app.

>   * Should we rename **Getting Started** to something like **Example code**? (Btw, I love that you have so much example code!)

**Getting Started** is more than just **Example code**, due to the speciality of the plugin, there are a lot of initial setup need to be done in the GooglePlay console and App Store Connect before the code can work. 

> * The purpose of the **Development** section is unclear.
>   
>   * Who is the first paragraph aimed at? A user of the API or a contributor to the package?

I renamed the **Development** to **Contributing to this plugin** so the purpose might be a little more clear? It is for contributors. 

>   * The second paragraph (about this being a beta plugin) seems pretty important. Should it be up at the top of the README?

I think you are right. It can be a second paragraph in the section # In App Purchase, right before ## Features. 

> * We need to establish some terminology. (It's fine to mention all the terms people might expect, but after the main term is established we should stick with it.) Which term do you want to use from the following groups?
>   
>   * _in-app purchase_ or _IAP_?

in-app purchase

>   * _storefront_, _underlying shop_, or _underlying store_?

underlying store .

**Comment by cyanglaz** [comment]
@kwalrath friendly ping regarding my last response :)

**Comment by kwalrath** [comment]
The gifs seem really large — both in number of bytes and in visual size. @jayoung-lee & @filiph, what do you think a good size would be?

File: packages/in_app_purchase/in_app_purchase/README.md
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase/README.md:6]: @jayoung-lee I have added videos here, the url is set on master, which will start work after landing the PR. 

To test it, check out my repo locally, update the url to `https://github.com/flutter/cyanglaz/blob/iap_readme/packages/in_app_purchase/in_app_purchase/doc/iap_android.gif?raw=true`, and view the readme locally.

**Comment by cyanglaz** [comment]
> The gifs seem really large — both in number of bytes and in visual size. @jayoung-lee & @filiph, what do you think a good size would be?

I manually set the height to be 400 in code, it looks ok to me, if you guys like it, I can update the gif file so it is smaller by default.

**Comment by kwalrath** [comment]
> > The gifs seem really large — both in number of bytes and in visual size. @jayoung-lee & @filiph, what do you think a good size would be?
> 
> I manually set the height to be 400 in code, it looks ok to me, if you guys like it, I can update the gif file so it is smaller by default.

400 px high lgtm

**Comment by kwalrath** [comment]
I'm working on a new version of the top README and have a couple of questions:

1. Should the API links be to API docs (e.g. [in_app_purchase](https://pub.dev/documentation/in_app_purchase/latest/in_app_purchase/in_app_purchase-library.html), [billing_client_wrappers](https://pub.dev/documentation/in_app_purchase/latest/billing_client_wrappers/billing_client_wrappers-library.html), and [store_kit_wrappers](https://pub.dev/documentation/in_app_purchase/latest/store_kit_wrappers/store_kit_wrappers-library.html)) instead of to .dart files?
2. The README refers to `PlayStore`, but I don't see that anywhere in the API.

**Comment by cyanglaz** [comment]
> I'm working on a new version of the top README and have a couple of questions:
> 
> 1. Should the API links be to API docs (e.g. [in_app_purchase](https://pub.dev/documentation/in_app_purchase/latest/in_app_purchase/in_app_purchase-library.html), [billing_client_wrappers](https://pub.dev/documentation/in_app_purchase/latest/billing_client_wrappers/billing_client_wrappers-library.html), and [store_kit_wrappers](https://pub.dev/documentation/in_app_purchase/latest/store_kit_wrappers/store_kit_wrappers-library.html)) instead of to .dart files?

links to API docs sounds good to me. 

> 2. The README refers to `PlayStore`, but I don't see that anywhere in the API.

PlayStore is the PlayStore on android phones: https://play.google.com/store?&utm_source=na_Med&utm_medium=hasem&utm_content=Nov0520&utm_campaign=Evergreen&pcampaignid=MKT-EDR-na-us-1000189-Med-hasem-py-Evergreen-Nov0520-Text_Search_BKWS-id_100754_%7cEXA%7cONSEM_kwid_43700045371544955&gclid=Cj0KCQjw9YWDBhDyARIsADt6sGZgFpiy5Zd-iRI1RTkVRhavG4EWAG1tuSUQ8qOmbcRR-bUIxqwyJWYaAnQcEALw_wcB&gclsrc=aw.ds

**Comment by kwalrath** [comment]
> PlayStore is the PlayStore on android phones: https://play.google.com/store

`PlayStore` is code fonted like it's an API, but I don't see it in the source code anywhere. I do see `IAPSource.GooglePlay` in the source code, but since `IAPSource` isn't shown in the code, it doesn't seem necessary to refer to the `GooglePlay` (or `AppStore`) identifier in the text. I'll just change those to the store names.

File: packages/in_app_purchase/README.md
Comment (jayoung-lee) [reply packages/in_app_purchase/README.md]: SG!

File: packages/in_app_purchase/in_app_purchase/README.md
Comment (jayoung-lee) [reply packages/in_app_purchase/in_app_purchase/README.md]: I could not access the local repo with the URL. Is the repo private?

**Comment by kwalrath** [comment]
I think we just need to fix the _listenToPurchaseUpdated/_handlePurchaseUpdates code and remove that PENDING note, and then this PR is good to go!

**Comment by cyanglaz** [comment]
> _listenToPurchaseUpdated

Done!

**Comment by cyanglaz** [comment]
@kwalrath @jayoung-lee Thanks for helping out, this looks pretty good now, could you please take a final review?

**Comment by kwalrath** [comment]
@cyanglaz it looks like the "[PENDING: We already showed code like this above (except the handler was called _listenToPurchaseUpdated). Copy this up there and remove from here?]" text is still there, and it hasn't been completely addressed. The code is still in two places, with different handlers (e.g. onDone & onError are only in the second example), so it's a bit confusing what exactly your code should look like.

**Comment by cyanglaz** [comment]
@kwalrath Sorry I missed that part. I copied the second one to the top. Fixed!

File: packages/in_app_purchase/in_app_purchase/README.md
Comment (kwalrath) [inline packages/in_app_purchase/in_app_purchase/README.md:62]:
```suggestion
* [Listening to purchase updates](#listening-to-purchase-updates)
* [Connecting to the underlying store](#connecting-to-the-underlying-store)
* [Loading products for sale](#loading-products-for-sale)
* [Loading previous purchases](#loading-previous-purchases)
```

File: packages/in_app_purchase/in_app_purchase/README.md
Comment (kwalrath) [inline packages/in_app_purchase/in_app_purchase/README.md:27]:
```suggestion
* Show in-app products that are available for sale from the underlying store.
   Products can include consumables, permanent upgrades, and subscriptions.
* Load in-app products that the user owns.
* Send the user to the underlying store to purchase products.
* Present a UI for redeeming subscription offer codes. (iOS 14 only)
```

**Review by kwalrath — APPROVED** [review]
lgtm once you fix the order of the usage toc and change that numbered list to a bulleted one (since it isn't a sequence: https://developers.google.com/style/highlights#formatting,-punctuation,-and-organization). Can't believe I didn't notice that before!

**Pull request merged**
